File: Algo/Leetcode/021MergeTwoSortedLinkedList.cpp
//1.Directly Compute. O(m+n)
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* res = head;
        
        while(l1&&l2){
            if(l1->val <= l2->val){head->next = l1;head = l1; l1 = l1->next;}
            else {head->next = l2;head = l2; l2 = l2->next;}
        }
        
        //if(!l1 && !l2) head->next=NULL;
        //else if(!l1) head->next=l2;
        //else if(!l2) head->next=l1;
        //else{}
        head->next = l1 ? l1 : l2;
        
        return res->next;
    }
};

 
//2. recursion O(m+n)
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {      
        if(!l1 && !l2) return NULL;
        else if(!l1) return l2;
        else if(!l2) return l1;
        else{
            if(l1->val <= l2->val) {l1->next=mergeTwoLists(l1->next,l2);return l1;}
            else {l2->next=mergeTwoLists(l1,l2->next);return l2;}
        }
    }
};

File: Algo/Leetcode/021MergeTwoSortedLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        head = ListNode()
        res = head
        while list1 or list2:
            if list1 and list2:
                if list1.val <= list2.val:
                    head.next = ListNode(list1.val)
                    list1 = list1.next
                else:
                    head.next = ListNode(list2.val)
                    list2 = list2.next
            elif list1 is None:
                head.next = ListNode(list2.val)
                list2 = list2.next
            else:
                head.next = ListNode(list1.val)
                list1 = list1.next

            head = head.next

        return res.next


# recursive
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        head = ListNode()
        if list1 and list2:
            if list1.val <= list2.val:
                head.val = list1.val
                list1 = list1.next
            else:
                head.val = list2.val
                list2 = list2.next
            head.next = self.mergeTwoLists(list1, list2)
        elif list1 is None:
            return list2
        else:
            return list1

        return head